src/particles/ImpactXParticleContainer.cpp

```diff
@@ -141,25 +141,52 @@ namespace impactx
                 gid = *it;
             }
         }
-        auto& particle_tile = DefineAndReturnParticleTile(lid, gid, tid);
 
-        auto old_np = particle_tile.numParticles();
-        auto new_np = old_np + np;
-        particle_tile.resize(new_np);
-
-        // Update NextID to include particles created in this function
-        int pid;
-#ifdef AMREX_USE_OMP
-#pragma omp critical (add_beam_nextid)
+    int nthreads = 1;
+#if defined(AMREX_USE_OMP)
+    nthreads = omp_get_max_threads();
 #endif
-        {
-            pid = ParticleType::NextID();
-            ParticleType::NextID(pid+np);
-        }
+
+    const auto& ba = ParticleBoxArray(lid);
+    auto n_logical =  numTilesInBox(ba[gid], true, tile_size);
+
+    if (n_logical < nthreads ) {
+        amrex::Print() << "Too few tiles for the number of OpenMP threads. Parallelization will be poor. \n";
+    }
+
+    for (int ithr = 0; ithr < nthreads; ++ithr) {
+        DefineAndReturnParticleTile(lid, gid, ithr);
+    }
+
+    int pid = ParticleType::NextID();
+    ParticleType::NextID(pid+np);
         AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
         static_cast<amrex::Long>(pid) + static_cast<amrex::Long>(np) < amrex::LongParticleIds::LastParticleID,
             "ERROR: overflow on particle id numbers");
 
+#if defined(AMREX_USE_OMP)
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+#endif
+    {
+        int tid = omp_get_thread_num();
+        int nr = np / nthreads;
+        int nlft = np - nr*nthreads;
+
+        int num_to_add = 0;
+        int my_index =0;
+        if (tid < nlft) { // get nr+1 items
+        my_index = tid * (nr+1);
+        num_to_add = nr+1;
+        } else {         // get nr items
```

Comment:
Can you please add inline comments on the meaning of the temporary variables (tid, nr, nlft, my_index), maybe spell them out for clarity, and a bit more details on the if/else? :pray: